Fix LDAP authentication when using StartTLS

smortex opened the pull request:
1. __Category:__ Bug fix
 
 2. __Github Issue:__
 
  - #985 (same symptoms)

 3. __Description of changes:__

When authenticating users against an LDAP directory that demand StartTLS
connections, the security plugin does not issue a StartTLS command to
upgrade the connection, leading to authentication failures.

The code use the ldaptive library to communicate with LDAP, but does not
use the same API as the one shown in the official ldaptive
documentation.  When used against an LDAP directory that demand
StartTLS, the examples of ldaptive perform the StartTLS upgrade
transparently and the communication with the directory is successful.

Update the code to use the same ldaptive API as the one of the working
examples.  This unbreak LDAP authentification when StartTLS is required.

As a bonus, remove the passing of the now useless properties all around
the code.


 4. __Why these changes are required?__ 

LDAP authentication is currently broken.

 5. __What is the old behavior before changes and new behavior after changes?__

I could not authenticate using my LDAP credentials in spite of the correct information being configured.

After patching, authentication works as expected.

 6. __Testing done:__ (_Please provide details of testing done: Unit testing, integration testing and manual testing_)

I only tested manually using _simple authentication with a DN and a password_.
 
I used the follwoing configuration to test (base configuration in the 1.0.0 tarball, adjusted for my LDAP directory, the system's trusted CA have been copied in `config/ca-certificates.crt`, and `./securityadmin_demo.sh` has been run to apply these parameters):

```sh-session
vagrant@buster:~/opensearch-1.0.0$ grep -v '#' plugins/opensearch-security/securityconfig/config.yml
---
_meta:
  type: "config"
  config_version: 2
config:
  dynamic:
    http:
      anonymous_auth_enabled: false
      xff:
        enabled: false
    authc:
      basic_internal_auth_domain:
        description: "Authenticate via HTTP Basic against internal users database"
        http_enabled: true
        transport_enabled: true
        order: 1
        http_authenticator:
          type: basic
          challenge: true
        authentication_backend:
          type: intern
      ldap:
        description: "Authenticate via LDAP"
        http_enabled: true
        transport_enabled: true
        order: 2
        http_authenticator:
          type: basic
          challenge: false
        authentication_backend:
          type: ldap
          config:
            enable_ssl: false
            enable_start_tls: true
            enable_ssl_client_auth: false
            verify_hostnames: true
            hosts:
              - ldap.blogreen.org
            bind_dn: cn=opensearch,ou=services,dc=blogreen,dc=org
            password: [redacted]
            userbase: 'ou=people,dc=blogreen,dc=org'
            pemtrustedcas_filepath: ca-certificates.crt
            usersearch: '(uid={0})'
            username_attribute: uid
    authz:
      roles_from_myldap:
        description: "Authorize via LDAP or Active Directory"
        http_enabled: true
        transport_enabled: true
        authorization_backend:
          type: ldap
          config:
            enable_ssl: false
            enable_start_tls: true
            enable_ssl_client_auth: false
            verify_hostnames: true
            hosts:
              - ldap.blogreen.org
            bind_dn: cn=opensearch,ou=services,dc=blogreen,dc=org
            password: [redacted]
            rolebase: 'ou=groups,dc=blogreen,dc=org'
            rolesearch: '(uniqueMember={0})'
            pemtrustedcas_filepath: ca-certificates.crt
            userroleattribute: null
            userrolename: disabled
            rolename: cn
            resolve_nested_roles: true
            userbase: 'ou=people,dc=blogreen,dc=org'
            usersearch: '(uid={0})'
```

After rebuilding the module with this path; replacing OpenSearch's `plugins/opensearch-security/opensearch-security-1.0.0.0.jar` with the one built; and restarting OpenSearch, authentication against LDAP directory and user group memberships from LDAP are working as expected.
 
 7. __TO-DOs, if any:__

- [x] Test against an LDAP directory that allows unauthenticated binds;
- [x] Test SASL authentication using a certificate;
- [x] Ensure correct behavior when hostname verification is disabled.

 8. __Is it backport from main branch?__

No



By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

vrozov reviewed (approved):
LGTM. Tested with anonymous bind. Thank you for the contribution @smortex!

smortex commented:
@vrozov Cool!  I updated the state of the TODO list above.  On my side I changed the value of `hosts` to another name which resolve to the LDAP server, ensured authentication failed because of a name mismatch, toggled `enable_ssl_client_auth` and I can sign in again.  So this item is also checked!

I don't have clues for the SASL related authentication though :shrug:.

smortex commented:
I could finaly check that EXTERNAL SASL authentication is working as expected.

Just a matter of `olcAuthzRegexp`, `olcTLSVerifyClient`, and ensuring the CA that signed the certificate file is in `olcTLSCACertificateFile`.

No more TODOs :wink: :balloon: :tada:
